syntax = "proto2";

package exemplo;

// serve para ordens de compra, venda
// e cancelamento
message Oferta {
  required int32 quantidade = 1;
  required int32 preco = 2; // em centavos
}

message MsgNegociacao {
  required string ativo = 1;
  required Oferta oferta = 2;
}

message MsgInfo {
  required string ativo = 1;
  optional bool notificar = 2;
}

message MsgAutenticacao {
  required string usuario = 1;
  required string senha = 2;
}

message MsgComando {
   oneof msg {
     MsgNegociacao compra = 1;
     MsgNegociacao venda = 2;
     MsgNegociacao cancela_venda = 3;
     MsgNegociacao cancela_compra = 4;
     MsgInfo info = 5;
     MsgAutenticacao autenticacao = 6;
     bool saida = 7;
   }
}


// as informações sobre um ativo, que fazem parte da resposta
// ao comando para obter informações
message MsgNotificacaoInfo {
  required string ativo = 1;
  required int32 ultimo_preco = 2;
  repeated Oferta compras = 3;
  repeated Oferta vendas = 4;
}

// mensagens de notificacação
message MsgNotificacaoExec {
  required string ativo = 1;
  required int32 preco = 2;
  required int32 quantidade = 3;
  required int32 tipo = 4; // compra, venda ou cancelamento ??? compra = 1, venda = 2, cancel = 3
}

// Mensagem de resposta genérica
message MsgResp {
  required int32 status = 1; //sucesso 1 ou falha 0
  oneof notificacao {
    MsgNotificacaoInfo info = 2;
    MsgNotificacaoExec exec = 3;
  }
}
